### ISSUE TITLE
Load forecast severely under-predicts daytime peaks — misses 20 kW mid-day load, forecasts only 2.5 kW

### ISSUE THREAD

**Issue by purcell-lab** [code-imstillalive/nimbus#257]
## Summary

On v0.94.12, the nimbus-owned load forecast (`sensor.nimbus_sigen_plant_total_load_power_forecast`) severely under-predicts the actual daytime load profile. Over 7 days of history, actual midday consumption averages **8-11 kW** with p90 reaching **18 kW** and peaks at **20+ kW**, while the forecast is a nearly-flat **1.1-2.8 kW** band across the whole horizon.

**Impact on the solver**: the 24-hour dispatch plan on 2026-08-28 06:40 AEST forecasts total load of **40.9 kWh/day**. Actual consumption over the last 5 days: **58, 77, 85, 90, 123 kWh/day (mean 86.5 kWh)** — under-forecast by 30-70%.

## Evidence

Load profile comparison (hour-of-day) — actual over 7 days vs nimbus's 24h-ahead forecast:

| hour | actual mean | actual p90 | actual max | nimbus forecast |
|---|---|---|---|---|
| 00-05 | 1.2-1.5 kW | 1.6-2.1 kW | 2.3-6.8 kW | **1.32 kW ✓** |
| 08 | 3.8 kW | 7.9 kW | 14.6 kW | 1.98 kW |
| 10 | 5.3 kW | 11.1 kW | 12.5 kW | 2.58 kW |
| 11 | **8.3 kW** | 18.3 kW | 21.4 kW | 2.64 kW |
| 12 | **8.8 kW** | 18.5 kW | 21.3 kW | 2.64 kW |
| 13 | **10.2 kW** | 18.6 kW | 20.5 kW | 2.67 kW |
| 14 | **10.6 kW** | 18.3 kW | 20.4 kW | 2.47 kW |
| 15 | **10.2 kW** | 18.3 kW | 18.6 kW | 2.36 kW |
| 16-23 | 1.5-2.4 kW | 2.9-4.3 kW | 5-19 kW | 1.32-1.68 kW |

Daily energy actual vs forecast:

| date | actual kWh | forecast (next 24h) |
|---|---|---|
| 2026-08-23 | 76.8 | — |
| 2026-08-24 | 122.5 | — |
| 2026-08-25 | 58.6 | — |
| 2026-08-26 | 84.7 | — |
| 2026-08-27 | 89.9 | — |
| **2026-08-28 (upcoming)** | — | **40.9** |

The Sigen Power Metrics chart for 2026-08-28 confirms the pattern — load spikes to -15 kW (consumption side) across 11:00-15:00.

## Root cause hypothesis

The load forecast sensor exposes:

```
mode: unscheduled
model_trained_at: 2026-08-28T03:00:00
training_points: 568
training_span_days: 5.91
validation_mae:
  knn: 1.76 kW
  gbrt: 1.45 kW (selected)
  naive: 1.68 kW
resample_minutes: 15
```

With **only 5.9 days** of 15-min-resampled training data (568 points), the model has not seen enough variance in the daytime peak-load pattern to reproduce it. GBRT's MAE of 1.45 kW is barely better than naive persistence (1.68 kW). The forecast then extrapolates the overnight baseline flat, occasionally rising to ~2.6 kW around noon — completely missing the 8-11 kW daytime plateau visible in the actual data.

There's a large daytime load (probably pool pump/HVAC/EV charging — need to identify which) that runs on a schedule the model hasn't captured because it hasn't seen enough repetitions.

## Impact on solver decisions

Looking at the 2026-08-28 dispatch plan:

- Solar forecast: 93.2 kWh
- Load forecast: 40.9 kWh (under by ~45 kWh vs actual mean)
- Plan: export 27 kWh, charge battery 36.7 kWh, discharge 19.1 kWh, zero import

If actual load runs to the 86.5 kWh mean:
- The 45 kWh gap will consume solar that the plan wanted to export
- Or, if the battery was fully charged during the (under-forecast) low-load window, the mid-day peak will draw from the battery faster than planned, potentially forcing evening imports at 35 c/kWh

The solver is picking a strategy optimised for a house that consumes half its actual energy. **Even a "correct" battery dispatch given the assumed load can be economically wrong if the load itself is wrong.**

## Options for improvement

1. **Longer training window** — 5.9 days is not enough to capture weekly patterns. 30+ days would let it learn weekday vs weekend, whether the pool pump runs, etc.
2. **Explicit schedule-aware inputs** — if the big daytime load is a pool pump on a timer, expose that timer state as an input feature or add it as a scheduled load in the plan.
3. **Use validation MAE to widen intervals** — the forecast currently has narrow p10-p90 bands (±0.5 kW) that give the solver false precision. With MAE of 1.45 kW, the bands should be at least ±2-3 kW.
4. **Fall back to `naive` or hour-of-day median** when validation MAE is close to naive — GBRT is barely beating persistence and its confident-but-wrong flat forecast is arguably worse than a lookup of "what did this hour do yesterday?"

## Probe recipe

```python
# 7-day actual load
GET /api/history/period/{now-7d}?filter_entity_id=sensor.sigen_plant_consumed_power
    &minimal_response=true&no_attributes=true&end_time={now}

# Nimbus load forecast
GET /api/states/sensor.nimbus_sigen_plant_total_load_power_forecast
# Read attributes.forecast[] for full 4-day horizon

# Bucket both by hour-of-day and compare mean/median/p90 profiles
```

## Not blocking

The solver is running optimally against the load forecast it's given — this is a **forecast quality** issue, not a solver bug. Filing so we can improve the load-forecast pipeline. Happy to attach the diagnostic snapshot and chart (uploading now).

**Comment by purcell-lab** [comment]
## Attached artifacts

Both files uploaded to a public gist: https://gist.github.com/purcell-lab/70c2aea7f73c9169892134724f3e15e6

- **Full v0.94.12 diagnostic**: [nimbus_diag_2026-08-28_0640.json](https://gist.githubusercontent.com/purcell-lab/70c2aea7f73c9169892134724f3e15e6/raw/nimbus_diag_2026-08-28_0640.json) — 478 KB, includes `solver.forecast[]` (361 rows), `solver_config`, blend algorithm, cost breakdown
- **Load forecast comparison chart** (SVG):

![load forecast vs actual](https://gist.githubusercontent.com/purcell-lab/70c2aea7f73c9169892134724f3e15e6/raw/b7f5479ca442e8ecafd42bbbd8792e17bb4373d7/load_forecast_vs_actual.svg)

The chart shows the 7-day actual load profile (p10-p90 band + median + mean) against the current Nimbus forecast, plus daily-total bars vs the 40.9 kWh forecast. The 8-11 kW midday plateau in the actual profile is what the current forecast (flat ~2.5 kW) is missing.

**Comment by purcell-lab** [comment]
## Proposed solution: use HA long-term statistics instead of recorder history

The load-forecast model trains on only **5.9 days / 568 points** because the recorder database only retains raw states for a few days. HA has a separate, permanent statistics store that keeps hourly aggregates **indefinitely**. Confirmed 30 days of hourly load available on this install (screenshot: [monthly view](https://gist.githubusercontent.com/purcell-lab/70c2aea7f73c9169892134724f3e15e6/raw/1000022871.png) shows the full August profile). Long-term stats also unlock **weekly and seasonal patterns** the current 5.9-day model cannot see.

## API surface

HA exposes two paths for reading long-term stats:

### Path 1: REST-callable service (recommended for nimbus integration)

`POST /api/services/recorder/get_statistics?return_response=true` — returns data in the response body.

**Request:**
```json
{
  "statistic_ids": ["sensor.sigen_plant_consumed_power"],
  "start_time": "2026-07-28 00:00:00",
  "end_time": "2026-08-27 23:59:59",
  "period": "hour",
  "types": ["mean", "max", "min"]
}
```

Available `period` values: `5minute` (kept ~10 days), `hour` (kept forever), `day`, `week`, `month`, `year`.
Available `types`: `mean`, `min`, `max`, `sum` (for total_increasing sensors), `change`, `last_reset`, `state`.

**Response shape:**
```json
{
  "service_response": {
    "statistics": {
      "sensor.sigen_plant_consumed_power": [
        {"start": "2026-07-28T12:00:00+00:00",
         "end":   "2026-07-28T13:00:00+00:00",
         "min": 1.663, "max": 3.178, "mean": 2.038},
        ...
      ]
    }
  }
}
```

Verified on the live install — returned 720 hourly rows over 30 days in <1s.

### Path 2: WebSocket API (equivalent, if the integration prefers push semantics)

Message type: `recorder/statistics_during_period` with the same fields plus `types: ["mean"]` etc. Same data; WS is what the HA frontend uses.

## Evidence: 30-day vs 5.9-day training set

Hour-of-day means from 30-day LTS (this install):

| hour (AEST) | 30-day mean (LTS) | 30-day p90 | 5.9-day mean (current model) |
|---|---|---|---|
| 00-05 | 1.9-2.1 kW | 2.6-3.1 kW | 1.32 kW |
| 07 | 3.8 kW | 7.8 kW | 1.13 kW |
| 08-10 | 3.0-4.8 kW | 6.0-9.7 kW | 1.98-2.58 kW |
| **11** | **9.4 kW** | 18.0 kW | 2.64 kW |
| **12** | **10.2 kW** | 19.0 kW | 2.64 kW |
| **13** | **10.1 kW** | 19.0 kW | 2.67 kW |
| **14** | **9.3 kW** | 18.0 kW | 2.47 kW |
| **15** | **6.7 kW** | 12.8 kW | 2.36 kW |
| 16-23 | 1.9-3.4 kW | 3.5-5.7 kW | 1.32-1.68 kW |

The 30-day LTS produces essentially the same daytime peak pattern as the 7-day recorder view but with **more stable statistics** (30 samples per hour vs 7). Both prove the 10 kW midday plateau is a **persistent, repeatable pattern**, not a one-off.

## Implementation options

### Option A (minimum change): expand training window via LTS

Change the load-forecast trainer's ingest path from recorder-history to LTS `hour` period. Same feature-engineering, but with 30-90 days of training data. LTS returns hourly means, so the model would predict hourly means and interpolate to the solver's 5-min grid.

**Trade-off:** loses 5-min resolution in training. Hourly is fine for the ~1 hour smoothing scale of home loads, but sharp transitions (e.g. pool pump starting) get blurred.

### Option B (best of both): hybrid 5-min recorder + hourly LTS

- Recent 3-5 days at 5-min from recorder → captures sharp intra-hour dynamics
- 30-90 days at `hour` from LTS → captures weekly/seasonal patterns

Blend by weighting: recent data at full weight, older LTS data at lower weight but with more coverage. Model sees both current shape and long-term profile.

### Option C: hour-of-day+day-of-week lookup fallback

Bypass ML entirely for the load leg — read 30-day LTS hourly means, group by (hour, day-of-week), and use the median as the forecast. This is what a Kalman filter or naive baseline would give and it beats the current 1.45 kW MAE GBRT in this scenario (the LTS-derived 10 kW at 12:00 has MAE < 5 kW vs the current model's ~8 kW).

## Additional benefit: solar and price too

The same LTS API works for `sensor.solcast_pv_forecast_forecast_today`, `sensor.amber_express_amber_general_price`, `sensor.sigen_inverter_battery_state_of_charge`, and every other continuous sensor. The forecast quality gains from LTS aren't load-specific — they apply to any sensor whose recorder retention is shorter than useful training history.

## Suggested next step

If Option A is agreed as the direction, I can:
1. Enumerate all `sensor.*_forecast` sensors nimbus produces to see how many use recorder-history vs external sources
2. PR a helper `read_long_term_statistics(entity_id, days=30, period="hour")` in `nimbus_load/data.py` that wraps the REST call
3. Wire the load-forecast trainer to use it, with a config option `training_source: recorder|lts|hybrid`

Happy to draft the PR skeleton if useful.

**Comment by purcell-lab** [comment]
Correction: the [monthly view] image reference in my previous comment doesn't resolve — I couldn't attach the user's uploaded History screenshot to a public location without web-UI upload. The **30-day hourly LTS table** in that comment is the substantive evidence and was pulled directly from the live install via `recorder.get_statistics`, not from the screenshot. The 27+ day span with the 20+ kW daytime spikes was independently confirmed from LTS.

**Comment by purcell-lab** [comment]
Draft PR is up: #258. Adds `training_source` option (recorder / lts / hybrid), keeps `recorder` as the default so existing installs are byte-identical on upgrade, binary sensors always force recorder. 8 new regression tests in `tests/test_coordinator_lts_hybrid_training.py`, all passing locally. Happy to iterate on the config-flow labels or the hybrid boundary math if you want a different default cutoff.
